lib/Transforms/HandshakeOptimizeBitwidths.cpp

```diff
@@ -657,23 +660,34 @@ struct HandshakeMuxSelect : public OpRewritePattern<handshake::MuxOp> {
     if (optWidth >= selectWidth)
       return failure();
 
-    // Replace the select operand with one with optimize bitwidth
-    Value newSelect =
-        modVal({selectOperand, ExtType::LOGICAL}, optWidth, rewriter);
-    rewriter.replaceUsesWithIf(selectOperand, newSelect, [&](OpOperand &oprd) {
-      return oprd.getOwner() == muxOp;
-    });
+    // Create a new mux whose select operand is optimized
+    SmallVector<Value, 3> newOperands;
+    newOperands.push_back(
+        modVal({selectOperand, ExtType::LOGICAL}, optWidth, rewriter));
+    auto dataOprds = muxOp.getDataOperands();
+    newOperands.append(dataOprds.begin(), dataOprds.end());
+    auto newMuxOp = rewriter.create<handshake::MuxOp>(
+        muxOp.getLoc(), muxOp->getResultTypes(), newOperands,
+        muxOp->getAttrs());
+    namer.replaceOp(muxOp, newMuxOp);
+    rewriter.replaceOp(muxOp, newMuxOp);
```

Comment:
The mux/cmerge rewrite patterns were already working as far as I can see. I just took the opportunity of the memory-related pattern refactorings to give them a fresh coat of paint. One minor difference is that the optimized mux/cmerge operations now keep the name of the operation they are replacing (c.f. `namer.replaceOp(muxOp, newMuxOp)` here). This is something I try to do now to maintain the traceability of semantically identical operations throughout the lowering (some people care).

This isn't the ideal way to do this because both the `namer` and the `rewriter` have to been made aware of the replacement, leaving a good chance that someone will forget one of the two at some point. I think the correct way to do this would be to register a "namer listener" on the pattern rewriter itself, so that calling `rewriter.replaceOp` automatically causes the naming analysis to update its name mapping. I hope to get the time to implement this correctly at some point.